Improve theme configuration load inside ThemeScript component

raffaelecarelle opened the pull request:
### 1. Why is this change necessary?
This changes fix use case which child theme script doesn't not load correctly

### 2. What does this change do, exactly?
Use AbstractConfigLoader component to load configuration

### 3. Describe each step to reproduce the issue or behaviour.
Create a custom theme
create a child theme of custom theme created above
create sales channel with child theme associated.

Now in the storefront scripts js doesn't load correctly (es. menu doesn't open or dropdowns don't work properly)

With styles/css there aren't issues

raffaelecarelle committed:
Improve theme configuration load using ConfigLoader component in ThemeScripts::getThemeScripts()

raffaelecarelle committed:
add changelog

raffaelecarelle committed:
fix changelog

Phil23 commented:
Hey @raffaelecarelle,

Your change seems to break 404 error page templates. You can test this manually by navigating to Settings -> Basic information and then setting a layout for 404 pages and then try to access any non existing route in the storefront. 

You can also test this with our automated test in the e2e test suite. It is the test `edit-information.cy.js` which is failing.

I could reproduce this on my local machine.

raffaelecarelle committed:
Refactor to use SalesChannelContext in ThemeScripts

This change updates the ThemeScripts class to use SalesChannelContext instead of Context. This ensures that the correct sales channel context is passed when loading theme configurations. Additionally, the sales channel context object is now retrieved from the request attributes to align with other system operations.

raffaelecarelle commented:
> Hey @raffaelecarelle,
> 
> Your change seems to break 404 error page templates. You can test this manually by navigating to Settings -> Basic information and then setting a layout for 404 pages and then try to access any non existing route in the storefront.
> 
> You can also test this with our automated test in the e2e test suite. It is the test `edit-information.cy.js` which is failing.
> 
> I could reproduce this on my local machine.

@Phil23  You were right. I fixed it and now e2e test passing

raffaelecarelle committed:
Add SalesChannelContext mocks in ThemeScriptsTest

Updated ThemeScriptsTest to include mock objects for SalesChannelContext. This ensures SalesChannelContext attributes are set correctly, enhancing test accuracy and coverage.

mitelg commented on changelog/_unreleased/2024-10-15-theme-script-with-configuration-load-improved.md:
```diff
@@ -0,0 +1,9 @@
+author_github: @raffaelecarelle
+---
+# Storefront
+* Changed `ThemeScript` component to use `AbstractConfigLoader`  to load theme configuration
```
we should write the  FQCN and not just the name of the class. that will avoid confusion

mitelg commented on src/Storefront/Theme/ThemeScripts.php:
```diff
@@ -50,12 +53,11 @@ public function getThemeScripts(): array
-        return $this->cache->get('theme_scripts_' . $path, function (ItemInterface $item) use ($themeName) {
-            $themeConfig = $this->pluginRegistry->getConfigurations()->getByTechnicalName($themeName);
+        /** @var SalesChannelContext $salesChannelContext */
+        $salesChannelContext = $request->attributes->get(PlatformRequest::ATTRIBUTE_SALES_CHANNEL_CONTEXT_OBJECT);
```
```suggestion:-1+0
        $salesChannelContext = $request->attributes->get(PlatformRequest::ATTRIBUTE_SALES_CHANNEL_CONTEXT_OBJECT);
        if (!$salesChannelContext instanceof SalesChannelContext) {
            return [];
        }
```

var annotations should not be used, as they could lie to PHPStan :sweat_smile: see https://github.com/shopware/shopware/blob/trunk/coding-guidelines/core/writing-code-for-static-analysis.md

mitelg commented on tests/unit/Storefront/Theme/ThemeScriptsTest.php:
```diff
@@ -81,6 +87,11 @@ public function testLoadPaths(): void
         $request->attributes->set(PlatformRequest::ATTRIBUTE_SALES_CHANNEL_ID, 'Storefront');
         $request->attributes->set(SalesChannelRequest::ATTRIBUTE_THEME_ID, 'Storefront');
         $request->attributes->set(SalesChannelRequest::ATTRIBUTE_THEME_NAME, 'Storefront');
+        $request->attributes->set(PlatformRequest::ATTRIBUTE_CONTEXT_OBJECT, Context::createCLIContext());
```
is that needed at all? :thinking: I see no usage of this one request attribute

mitelg commented on tests/unit/Storefront/Theme/ThemeScriptsTest.php:
```diff
@@ -81,6 +87,11 @@ public function testLoadPaths(): void
         $request->attributes->set(SalesChannelRequest::ATTRIBUTE_THEME_ID, 'Storefront');
         $request->attributes->set(SalesChannelRequest::ATTRIBUTE_THEME_NAME, 'Storefront');
+        $request->attributes->set(PlatformRequest::ATTRIBUTE_CONTEXT_OBJECT, Context::createCLIContext());
+        $salesChannelContext = $this->createMock(SalesChannelContext::class);
```
instead of creating a mock, this helper could also be used: `\Shopware\Core\Test\Generator::createSalesChannelContext`

mitelg commented on tests/unit/Storefront/Theme/ThemeScriptsTest.php:
```diff
@@ -119,6 +131,12 @@ public function testInheritsFromBase(): void
+
+        $salesChannelContext = $this->createMock(SalesChannelContext::class);
+        $salesChannelContext->method('getContext')->willReturn(Context::createCLIContext());
+        $request->attributes->set(PlatformRequest::ATTRIBUTE_SALES_CHANNEL_CONTEXT_OBJECT, $salesChannelContext);
```
yes, the test is running again, with all those additions, but it is not really ensured, that the described problem is fixed

raffaelecarelle replied:
Yes it is. I change attributes of "testInheritsFromBase", theme child can't it is parent itself

mitelg replied:
looks better now :+1:

raffaelecarelle committed:
use  FQCN instead

raffaelecarelle committed:
Prevent null sales channel context in ThemeScripts

Add check to ensure $salesChannelContext is an instance of SalesChannelContext to prevent potential errors. This avoids issues when the sales channel context attribute is not properly set.

raffaelecarelle committed:
Refactor context object assignment in ThemeScriptsTest

Remove redundant context object assignment to streamline the test setup. This change simplifies the initialization of the necessary attributes for the request in the ThemeScriptsTest class.

raffaelecarelle committed:
Refactor sales channel context creation in tests

Replace manual mock creation of SalesChannelContext with the Generator::createSalesChannelContext() method. This improves code maintainability and reduces redundancy across test cases.

raffaelecarelle committed:
Update test theme attributes in ThemeScriptsTest

Refactor theme ID and name to use 'ChildId' and 'ChildName' instead of 'Storefront' and null. This change ensures consistency and correctness in the request attributes during the test.

raffaelecarelle committed:
Remove unused SalesChannelContext import

The SalesChannelContext import was identified as unused in ThemeScriptsTest.php. Removing it helps to keep the code clean and maintainable. No functionality is impacted by this change.

Phil23 commented:
@raffaelecarelle Thank you for your contribution. The changes are merged now. 👍
